# 0.1 Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 0.2 Importing Dataframes

In [2]:
# Set filepath for IC Basket project
path = r'/Users/joefeuille/CF Analysis/Instacart Basket Analysis'

In [3]:
path

'/Users/joefeuille/CF Analysis/Instacart Basket Analysis'

In [4]:
# Set Orders set as df_ords
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [5]:
# Set Products set as df_prods
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 0.3 Data Wrangling

### Dropping Columns

In [6]:
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [7]:
df_ords['eval_set'].value_counts(dropna=False)

prior    3214874
train     131209
test       75000
Name: eval_set, dtype: int64

In [8]:
df_ords['days_since_prior_order'].value_counts(dropna=False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [9]:
df_ords.value_counts()

order_id  user_id  eval_set  order_number  order_dow  order_hour_of_day  days_since_prior_order
1         112108   train     4             4          10                 9.0                       1
2280953   767      prior     13            1          18                 23.0                      1
2280955   119072   prior     10            5          13                 29.0                      1
2280956   135372   prior     72            5          18                 3.0                       1
2280957   73662    prior     6             1          13                 8.0                       1
                                                                                                  ..
1140724   131093   prior     21            1          16                 6.0                       1
1140725   85700    prior     21            2          11                 6.0                       1
1140726   201418   prior     25            6          10                 3.0                    

In [10]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [11]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [12]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [13]:
df_ords.describe()

user_id  order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.029782e+05  1.715486e+01        2.776219e+00       1.345202e+01   
std    5.953372e+04  1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00  1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.139400e+04  5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.026890e+05  1.100000e+01        3.000000e+00       1.300000e+01   
75%    1.543850e+05  2.300000e+01        5.000000e+00       1.600000e+01   
max    2.062090e+05  1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [14]:
df_ords['order_id'].dtype

dtype('O')

# Importing new dataframe - departments

In [15]:
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [16]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [17]:
# Transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [18]:
df_dep_t = df_dep.T

In [19]:
df_dep_t.head()

0
department_id  department
1                  frozen
2                   other
3                  bakery
4                 produce

In [20]:
# Cleaning up header with an index reset
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [21]:
# Removing first row to make a new header
new_header = df_dep_t.iloc[0]

In [22]:
new_header

0    department
Name: department_id, dtype: object

In [23]:
df_dep_t_new = df_dep_t[1:]

In [24]:
df_dep_t_new.head()

0
1   frozen
2    other
3   bakery
4  produce
5  alcohol

In [25]:
#set the header row as the df header
df_dep_t_new.columns = new_header

In [26]:
df_dep_t_new.head()

department_id department
1                 frozen
2                  other
3                 bakery
4                produce
5                alcohol

# Creating a Data Dictionary

In [27]:
data_dict = df_dep_t_new.to_dict('index')

In [28]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [29]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [30]:
print(data_dict.get('19'))

{'department': 'snacks'}


# Subsetting & Indexes

In [31]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [32]:
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [33]:
df_snacks = df_prods[df_prods['department_id']==19]

In [34]:
df_snacks.head(10)

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   
55          56                   Healthy Pop Butter Popcorn        23   
58          59         Medium Taqueria Style Chipotle Salsa        50   
76          77            Coconut Chocolate Chip Energy Bar         3   
77          78            Nutter Butter Cookie Bites Go-Pak        61   
92          93                  Uncured Cracked Pepper Beef        23   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5  
55             19     4.4  
58             19     6.8  
76             19     6.4  
77             19     3.0  
92             19     2.4

# Task 4.4

## Q.2: Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [35]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(4), object(2)
memory usage: 182.7+ MB


In [36]:
# Changing user_id variable from integer to string/object
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [37]:
# Varifying user_id was changed correctly
df_ords['user_id'].dtype

dtype('O')

## Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the data frame.

In [38]:
# Copying to new DF
df_ords_new = df_ords

In [39]:
# Renaming order_hour_of_day to order_time_of_day
df_ords_new.rename(columns = {'order_hour_of_day':'order_time_of_day'},
                    inplace = True)

In [40]:
df_ords_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 object 
 2   eval_set                object 
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_time_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(3), object(3)
memory usage: 182.7+ MB


## Q.4: Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [41]:
df_ords_new['order_time_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_time_of_day, dtype: int64

#### The busiest time of day seems to be 9-17, which is 9am to 5pm.

## Q.5: Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [42]:
# Checking our data dictionary
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [43]:
df_prods.head(10)

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   
5           6                                       Dry Nose Oil        11   
6           7                     Pure Coconut Water With Orange        98   
7           8                  Cut Russet Potatoes Steam N' Mash       116   
8           9                  Light Strawberry Blueberry Yogurt       120   
9          10     Sparkling Orange Juice & Prickly Pear Beverage       115   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3  
5             11     2.6  
6              7     4.4  
7              1     1.1  
8             16     7.0  
9              7     8.4

In [44]:
df_produce = df_prods[df_prods['department_id']==4]

In [45]:
df_produce.head()

product_id                    product_name  aisle_id  department_id  \
30          31              White Pearl Onions       123              4   
42          43             Organic Clementines       123              4   
44          45               European Cucumber        83              4   
65          66       European Style Spring Mix       123              4   
88          89  Yogurt Fruit Dip Sliced Apples       123              4   

    prices  
30     7.5  
42    11.5  
44    14.3  
65    11.7  
88    12.6

## Q.6: The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [46]:
# Creating new dataframe for breakfast items:
# breakfast (14), dairy eggs (16), & bakery(3)
df_brkf = df_prods.loc[df_prods['department_id'].isin([3,14,16])]

In [47]:
df_brkf.head(10)

product_id                       product_name  aisle_id  department_id  \
8             9  Light Strawberry Blueberry Yogurt       120             16   
27           28                  Wheat Chex Cereal       121             14   
33           34                                NaN       121             14   
57           58              Whole Wheat Tortillas       128              3   
67           68            Pancake Mix, Buttermilk       130             14   
82           83        100% Whole Wheat Pita Bread       128              3   
89           90                       Smorz Cereal       121             14   
97           98          2% Yellow American Cheese         2             16   
100         101         Bread, Healthy Whole Grain       112              3   
107         108      Organic Sprouted Barley Bread       112              3   

     prices  
8       7.0  
27     10.1  
33     12.2  
57      6.7  
67     13.7  
82     14.4  
89      3.9  
97      4.5  
100     5.2  
107    14.4

## Q.7: Find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [48]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [49]:
# Creating new dataframe for alcohol (5), deli (20), bev (7),
# and meat/seaf (12)
df_din_party = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [50]:
df_din_party.head(10)

product_id                                       product_name  aisle_id  \
2            3               Robust Golden Unsweetened Oolong Tea        94   
6            7                     Pure Coconut Water With Orange        98   
9           10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                                  Peach Mango Juice        31   
16          17                                  Rendered Duck Fat        35   
19          20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
22          23                             Organic Turkey Burgers        49   
34          35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
38          39           Daily Tangerine Citrus Flavored Beverage        64   
39          40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1  
19              7     6.0  
22             12     8.2  
34             12    15.1  
38              7    12.5  
39             12    22.5

## Q.8: It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [51]:
df_din_party.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7650 entries, 2 to 49688
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     7650 non-null   int64  
 1   product_name   7647 non-null   object 
 2   aisle_id       7650 non-null   int64  
 3   department_id  7650 non-null   int64  
 4   prices         7650 non-null   float64
dtypes: float64(1), int64(3), object(1)
memory usage: 358.6+ KB


### df_din_party has 7650 rows of data and 5 columns.

## Q.9:  Extract all the information you can about this user with "user_id" of "1".

In [52]:
df_user1 = df_ords.loc[df_ords['user_id'].isin(['1'])]

In [53]:
df_user1.head(20)

order_id user_id eval_set  order_number  orders_day_of_week  \
0   2539329       1    prior             1                   2   
1   2398795       1    prior             2                   3   
2    473747       1    prior             3                   3   
3   2254736       1    prior             4                   4   
4    431534       1    prior             5                   4   
5   3367565       1    prior             6                   2   
6    550135       1    prior             7                   1   
7   3108588       1    prior             8                   1   
8   2295261       1    prior             9                   1   
9   2550362       1    prior            10                   4   
10  1187899       1    train            11                   4   

    order_time_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

## Q.10: You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [54]:
df_user1.describe()

order_number  orders_day_of_week  order_time_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

## Q.11: Check the organization and structure of your notebook. Be sure to include section headings and code comments.

### Done.

## Q.12: Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [55]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [56]:
# Exporting df_ords_new from Q.3 too.
df_ords_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled2.csv'))

## Q.13: Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [57]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))